### PR TITLE
windows: Support EMIT_X64/X86 with msvc toolchain

### PR DESCRIPTION
**Pull request by jwang36**
The patches in this PR add support of MICROPY_EMIT_X64 and MICROPY_EMIT_X86 for msvc toolchain in ports/windows. They are enabled by default if building with msvc toolchain, but not with mingw and cygwin toolchain. It includes following changes:

- Add Windows X64 calling convention
- Add assembly code (masm) to support nlr_push and nlr_jump because msvc doesn't support inline assembly for X64 code
- Add alloc.c and gccollect.c (tweaked from unix/alloc.c and unix/gccollect.c) to use Windows memory management for gc

The patches pass all test cases of tests/micropython/native_*,viper_* on Windows.

### CODE REVIEW

File: ports/windows/alloc.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/*
```
Comment (stinos) [inline ports/windows/alloc.c]: This looks like a copy of unix/alloc.c with like 3 lines changed. In such case I think it's better to modify unix/alloc.c instead to make it work with windows builds.

**Comment by stinos** [comment]
Interesting, but not all Appveyor builds pass: the native_try/viper_try tests fail for the 64bit debug build

File: ports/windows/gccollect.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/*
```
Comment (stinos) [inline ports/windows/gccollect.c]: Same remark as the other file: I'm not sure there even is a functional difference with unix/gccollect.c?

File: ports/windows/micropython.vcxproj [hunk 3]
```diff
@@ -82,6 +83,9 @@
   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
     <ClCompile />
     <Link />
+    <Link>
+      <ImageHasSafeExceptionHandlers>false</ImageHasSafeExceptionHandlers>
```
Comment (stinos) [inline ports/windows/micropython.vcxproj]: Why is this needed, and why only for Release|Win32 ?

File: ports/windows/gccollect.c (hunk 2 above)
Comment (jwang36) [reply ports/windows/gccollect.c]: There's windows api used. I'm not sure it's ok to put them in unix folder. So I made a copy.

File: ports/windows/micropython.vcxproj (hunk 3 above)
Comment (jwang36) [reply ports/windows/micropython.vcxproj]: My fault. It's just for an irrelevant inspection for code generation in windows. I forgot to remove. Thanks.

**Comment by jwang36** [comment]
Any suggestion on the AppVeyor fails? I cannot reproduce it on my Windows 10.

File: ports/windows/msvc/asmnlrx64.asm [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    mov     [rcx + 20h], rbp
+    mov     [rcx + 28h], rdi
+    mov     [rcx + 30h], rsi
+    mov     [rcx + 38h], r12
+    mov     [rcx + 40h], r13
+    mov     [rcx + 48h], r14
+    mov     [rcx + 50h], r15
+    mov     [rcx + 58h], rdx   ; rip
+
+    push   rdx                      ;nlr_push_tail needs the return address
```
Comment (stinos) [inline ports/windows/msvc/asmnlrx64.asm]: I think the code is fine, but for the sake of consistency the comments should at least all lign up the same way, or else just be put right after the statements.

File: ports/windows/msvc/sources.props [hunk 5]
```diff
@@ -35,4 +34,16 @@
     <ClInclude Include="$(PyBaseDir)ports\windows\*.h" />
     <ClInclude Include="$(PyBaseDir)ports\windows\msvc\*.h" />
   </ItemGroup>
+  <ItemGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
```
Comment (stinos) [inline ports/windows/msvc/sources.props]: This can be simplified to oonly look at Platform for the condition, since the Configuration doesn't matter.

Comment (jwang36) [reply ports/windows/msvc/sources.props]: You're right. I'll remove the Configuration.

File: ports/windows/msvc/asmnlrx64.asm (hunk 4 above)
Comment (jwang36) [reply ports/windows/msvc/asmnlrx64.asm]: Agree. It will be aligned.

File: ports/windows/gccollect.c (hunk 2 above)
Comment (stinos) [reply ports/windows/gccollect.c]: If it's only a slight change, or a function which has the same signature in POSIX but just a different name, it's ok to do this I think. Else the way this has be fixed until now is to define our own implementation of the posix function (for example like is done for realpath/gettimeofday/opendir/...).

**Comment by stinos** [comment]
> Any suggestion on the AppVeyor fails? I cannot reproduce it on my Windows 10.

If you don't find an actual machine to reproduce it on, log in to Appveyor for your fork, adjust the settings to allow you to login with remote desktop (https://www.appveyor.com/docs/how-to/rdp-to-build-worker/) then start a build and login to the VM.

**Comment by stinos** [comment]
I was thinking it's probably better to split this in separate commits: one for the changes to nlr, one for the other changes, because those 2 are basically unrelated (as in, first one can exist without the other) and that weill be easier to see what goes on.  Maybe even a seperate commit for implementing alloc.c but that's maybe a bit too much. On the other hand it would still be easier to follow, imo.

Also since nlr_jump is now available in assembly, wouldn't it be possible to use it instead of setjmp for collecting regs/stack as well (i.e. MICROPY_GCREGS_SETJMP put to 0)?

**Comment by jwang36** [comment]
Ok. I'll first fix the failed cases, and then split the commits with other changes to address review comments got so far. Thanks

**Comment by dpgeorge** [comment]
Thanks for the contribution.  As @stinos says, it'd be best to not duplicate alloc.c/gccollect.c if they are basically the same as in the unix port.  The unix port already has some `_WIN32` specific code.

And please do separate the changes to asmx64/emitnx64 into a separate commit, since they are logically separate (adding WIN64 ABI support to this emitter).  If in doubt prefer more commits rather than less, since they can always be squashed together during merge if needed.

> not all Appveyor builds pass: the native_try/viper_try tests fail for the 64bit debug build

I think this test also fails on the unix debug build, maybe x86 (but that's not part of Travis).  I think it's related to `MP_OBJ_STOP_ITERATION` being different for debug vs non-debug builds.  The addition of `MICROPY_DEBUG_MP_OBJ_SENTINELS` was supposed to fix this though.

**Comment by dpgeorge** [comment]
Closing due to inactivity.  It's a nice feature so feel free to reopen and address review comments.
